allensdk/brain_observatory/behavior/behavior_session.py

```diff
@@ -4,29 +4,50 @@
 import numpy as np
 import inspect
 
+from allensdk.core.lazy_property import LazyPropertyMixin
 from allensdk.brain_observatory.behavior.session_apis.data_io import (
     BehaviorLimsApi, BehaviorNwbApi)
 from allensdk.brain_observatory.behavior.session_apis.abcs import BehaviorBase
+from allensdk.brain_observatory.behavior.trials_processing import (
+    calculate_reward_rate)
+from allensdk.brain_observatory.behavior.dprime import (
+    get_rolling_dprime, get_trial_count_corrected_false_alarm_rate,
+    get_trial_count_corrected_hit_rate)
+from allensdk.brain_observatory.behavior.dprime import (
+    get_hit_rate, get_false_alarm_rate)
+
 
 BehaviorDataApi = Type[BehaviorBase]
 
 
-class BehaviorSession(object):
+class BehaviorSession(LazyPropertyMixin):
     def __init__(self, api: Optional[BehaviorDataApi] = None):
```

Comment:
@njmei at a high level could you please describe what `LazyProperty` does?

Reply:
A `LazyProperty` based on my understanding, is kind of like a variable/attribute promise. When setting a variable or attribute, you provide a callable, which will be called (to provide data) when the variable/attribute __get__ method is called.

It's also used to 'mark' attributes that need to be compared with the `session_are_equal()` equality testing function.

Reply:
Weren't we already lazily loading data when the property is called?

I would much prefer to see a decorator such as 
```
@needs_equality_check
```

or similar.